### PR TITLE
Final algebra Semantics Blog Post

### PR DESCRIPTION
**Pull request by maxsnew**
Blog post based on reading @mwand 's final algebra paper and some discussions with @michaelballantyne and @rmculpepper

A couple things:
1. @michaelballantyne do you want me to link your blog in my post? It looks like you haven't updated it since 2014 and if not do you have a different website to link?
2. The part about adequacy can probably be expanded.

### CODE REVIEW

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+In the process we'll learn a bit about categorical logic, and I'll
+reformulate some of the category theory in that paper to be a bit
+more modern in presentation, cleaning some things up in the process.
+
+<!-- more -->
+
+# Intuition: Implementing a Signature
+
+As a running example, say we wanted to implement a datatype of finite
+maps whose keys and values are both integer, i.e., finite multisets
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: typo: should be "both integers"

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+We could specify such a datatype by specifying a little language of
+numbers and finite multisets.
+We'll have two "sorts" `num` and `multiset`, a constant for every integer, and an addition function
+```
+'n : () -> num
+add : (num, num) -> num
+```
+subject to the silly-looking equation:
+```
+add('n,'m) = n + m
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: I guess this should be `'(n + m)`...

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+empty : () -> multiset;
+singleton : (num) -> multiset;
+union : (multiset, multiset) -> multiset;
+remove : (num, multiset) -> multiset;
+count : (num, multiset) -> num
+```
+subject to the computational equations:
+
+```
+count('n, empty) = 0
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: `'0`

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+First, our little language of numbers and multisets is called a
+*theory*. The specific category gadget that we'll use to describe it
+is a *multi-sorted Lawvere theory*, or just *Lawvere theory* for
+short.
+
+A *Lawvere theory* is a category with finite products all of whose
+objects are finite products of a collection of *sorts* \\(S\\). We can
+construct this category from our little language above by making the
+objects be *contexts* \\(x:num,y:multiset,...\\) and morphisms \\(\Gamma ->
+x_1:s_1,...,x_n:s_n\\) to be \\(n\\)-tuples of terms \\(\Gamma -> s_1,...,
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: why "terms"?

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+First, our little language of numbers and multisets is called a
+*theory*. The specific category gadget that we'll use to describe it
+is a *multi-sorted Lawvere theory*, or just *Lawvere theory* for
+short.
+
+A *Lawvere theory* is a category with finite products all of whose
+objects are finite products of a collection of *sorts* \\(S\\). We can
+construct this category from our little language above by making the
+objects be *contexts* \\(x:num,y:multiset,...\\) and morphisms \\(\Gamma ->
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: what is `\Gamma`?

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md (hunk 4 above)
Comment (bennn) [reply blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: yeah I'm very confused, is `\Gamma -> s_1`  supposed to be an exponential object?

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+x_1:s_1,...,x_n:s_n\\) to be \\(n\\)-tuples of terms \\(\Gamma -> s_1,...,
+\Gamma -> s_n\\) *modulo* the equations we've specified. We'll use the
+letter \\(T\\) to mean a Lawvere theory.
+
+Then a *\\(T\\)-algebra* is a denotational semantics of our theory \\(T\\),
+i.e., a product preserving functor \\(A : T \to Set\\). This means for
+every sort we get a set \\(A(s)\\) and for every term \\(x_1:s_1,...,x_n:s_n
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: hmm so "sorts" and "terms" are 2 kinds of objects in `T`?

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+\vdash t : s\\) a function \\(A(t) : A(s_1)\times\cdots \times A(s_n) \to
+A(s)\\).
+
+Finally a *morphism of \\(T\\)-algebras* from \\(A\\) to \\(B\\) is a way to
+translate one algebra into another. Briefly, it is a natural
+transformation from \\(A\\) to \\(B\\), but concretely this means for every
+sort \\(s\\) we get a function \\(\alpha_s : A \to B\\) that translates \\(A\\)s
+interpretation into \\(B\\)s. The key property that we want is that the
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: suggestion: clarify "A's interpretation of `s` as a `Set`-object"

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+\\(In(T)(s) = \cdot \vdash s\\), the closed terms of that sort modulo the
+equivalence of the theory, and \\(In(T)(s_1,\ldots,s_n) =
+In(T)(s_1)\times\ldots,In(T)(s_n)\\). Then the terms are just
+interpreted as the functions you get by plugging closed inputs into
+them. Then if we look at what what a morphism of \\(T\\)-algebras from
+\\(In(T) \to A\\) is, we see that we don't have any choice, the only one
+is the one that maps \\(\cdot \vdash t : s\\) to \\(A(t)\\) and this makes all
+the right diagrams to commute.
+This is pretty similar to our definition of "initial algebra" before,
+except that `count` was defined as a function, not an ADT, but that
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: "was defined" => "is defined here as a function"

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+was just an easy way to satisfy the computational equations for
+`count`.
+
+However, an egregious flaw presents itself when we look at what the
+*final* algebra is. It's completely trivial! We can define \\(Fin(T)\\) to
+take every sort to a one element set \\(Fin(T)(s) = \{*\}\\) and every
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: typo: `|`

(was caps lock on?)

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+The strictness ensures that we don't mix up our sorts and our
+contexts, a morphim of theories has to map sorts to sorts, whereas a
+non-strict functor could map a sort to a context with two sorts it's
+equivalent to.
+What this really amounts to is a translation of one theory into
+another. It maps sorts to sorts and terms to terms of the appropriate
+sorts in a compositional way.
+However, we don't want to consider *all* such morphisms, only the ones
+that are "conservative extensions", i.e., there are no new closed
+terms at old types and closed terms that were different before remain
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: suggestion: "(1) there are no new closed terms at old types and (2) closed terms ...."

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+interpretation into \\(B\\)s. The key property that we want is that the
+operations according to \\(A\\) and \\(B\\) do the same thing as determined by
+\\(\alpha\\). Specifically, for any term \\(x_1:s_1,...,x_n:s_n \vdash t :
+s\\), and inputs \\(x_1 \in A(s_1),...,x_n \in A(s_n)\\) we should get the
+same result if we evaluate \\(A(t)(x_1,\ldots,x_n)\\) and then apply
+\\(\alpha_s\\) as if we first translate \\(x_1,\ldots,x_n\\) to
+\\(B(s_1),\ldots,B(s_n)\\) and then apply \\(B(t)\\). If you unwind the
+definitions, this is exactly what naturality says.
+
+Then we have a category of \\(T\\)-algebras and we can ask if there are
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: introduce the name `T-Alg` here

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+what it really ensures is that \\(A\\) determines a *congruence relation*
+on \\(T_1\\) given by \\(t1 \equiv_A t2\\) if \\(A(t1) = A(t2)\\).
+In light of this, the first condition could be called *adequacy*.
+
+Furthermore, the surjectivity condition ensures that any morphism of
+\\(i\\) algebras, i.e., a map as \\(T_1\\)-algebras is also surjective, so a
+morphism \\(A \to B\\) is a witness to the fact that \\(B\\) determines a
+*stronger* congruence relation on \\(T_1\\) than \\(A\\) does: \\(t1 \equiv_B t2
+\implies t1 \equiv_A t2\\).
+Then asking for a final algebra is asking for exactly the
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: the:

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+like domains or metric spaces or time-varying values.
+
+Another question is how to implement this when we have a proper *type
+theory* and not just some boring sorts. In particular, if we have
+function types, then we won't be able to get functions from functions
+in our term model to functions in our denotations due to
+contravariance. Perhaps logical relations are the solution?
+
+[ryanc]: http://ccs.neu.edu/home/ryanc
+[michaelb]: ???
```
Comment (bennn) [inline blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: maybe https://github.com/michaelballantyne

Comment (michaelballantyne) [reply blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: Yeah, my github is probably the best right now.

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md (hunk 5 above)
Comment (maxsnew) [reply blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: it's a context

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md (hunk 4 above)
Comment (maxsnew) [reply blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: I'll change it to `\Gamma \vdash t_1 : s_1`

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md (hunk 6 above)
Comment (maxsnew) [reply blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: sorts are the objects, terms are the arrows

File: blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md (hunk 9 above)
Comment (maxsnew) [reply blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md]: What's the problem here?

**Pull request merged**